### PR TITLE
fix(types/unstable): better type hints for Deno.dlopen (close: #16873)

### PR DESCRIPTION
**Pull request by ahaoboy**
close https://github.com/denoland/deno/issues/16873
Use magic code wrap parameters type and get more accurate return type
```
  type Cast<A, B> = A extends B ? A : B;
  type Const<T> = Cast<
    T,
    | (T extends string | number | bigint | boolean ? T : never)
    | { [K in keyof T]: Const<T[K]> }
    | []
  >;
declare function f<T>(arg: Const<T>): T;


let r = f({
  a: {
    b: 1,
    c: "2",
  },
});
```

image

### CODE REVIEW

File: cli/tsc/dts/lib.deno.unstable.d.ts [hunk 1]
```diff
@@ -611,7 +622,7 @@ declare namespace Deno {
    */
   export function dlopen<S extends ForeignLibraryInterface>(
     filename: string | URL,
-    symbols: S,
+    symbols: Const<S>,
   ): DynamicLibrary<S>;
```
Comment (aapoalas) [inline cli/tsc/dts/lib.deno.unstable.d.ts:626]: Note that you get the same hints with just marking the declaration object `as const`.

**Comment by littledivy** [comment]
Thank you for the PR @ahaoboy 

Can you add type tests in [`test_ffi/tests/ffi_types.ts`](https://github.com/denoland/deno/blob/main/test_ffi/tests/ffi_types.ts) ?

File: cli/tsc/dts/lib.deno.unstable.d.ts (hunk 1 above)
Comment (aapoalas) [reply cli/tsc/dts/lib.deno.unstable.d.ts:626]: Yeah, that's how I read it as well. It's neat of course :)

Comment (ahaoboy) [reply cli/tsc/dts/lib.deno.unstable.d.ts:626]: This trick often used in wasm/napi, because as const is redundant

**Comment by ahaoboy** [comment]
@littledivy 
I remove all as const in [test_ffi/tests/ffi_types.ts](https://github.com/denoland/deno/blob/main/test_ffi/tests/ffi_types.ts),   This tick  also works for both UnsafeCallback and UnsafeFnPointer.

**Review by littledivy — APPROVED** [review]

**Pull request merged**